pkg/certificate/configuration.go

```diff
@@ -2,169 +2,453 @@ package certificate
-	return webhook.(*admissionregistrationv1.ValidatingWebhookConfiguration)
+func (k objectKey) String() string {
+	return fmt.Sprintf("%s/%s", k.Kind, k.NamespacedName.String())
 }
 
-// clientConfigList returns the the list of webhooks's mutation or validating WebhookClientConfig
-//
-// The WebhookClientConfig type is share between mutating or validating so we can have a common function
-// that uses the interface client.Object and do some type checking to access it [1].
-//
-// [1] https://godoc.org/k8s.io/kubernetes/pkg/apis/admissionregistration#WebhookClientConfig
-func (m *Manager) clientConfigList(webhook client.Object) []*admissionregistrationv1.WebhookClientConfig {
-	clientConfigList := []*admissionregistrationv1.WebhookClientConfig{}
-	if m.webhookType == MutatingWebhook {
-		mutatingWebhookConfig := mutatingWebhookConfig(webhook)
-		for i, _ := range mutatingWebhookConfig.Webhooks {
-			clientConfig := &mutatingWebhookConfig.Webhooks[i].ClientConfig
-			clientConfigList = append(clientConfigList, clientConfig)
-		}
-	} else if m.webhookType == ValidatingWebhook {
-		validatingWebhookConfig := validatingWebhookConfig(webhook)
-		for i, _ := range validatingWebhookConfig.Webhooks {
-			clientConfig := &validatingWebhookConfig.Webhooks[i].ClientConfig
-			clientConfigList = append(clientConfigList, clientConfig)
-		}
+// keyedObject references a K8s resource object by key
+type keyedObject struct {
+	key     *objectKey
+	kobject client.Object
+}
+type objectMap map[*objectKey]*keyedObject
+
+// objectOperators defines functors for initializing and mapping k8s resources
+// to/from certificate chain data
+type objectOperators struct {
+	creator         func(name, namespace string) client.Object
+	toChainMapper   func(*keyedObject, objectMap, *chain.CertificateChainData)
+	fromChainMapper func(*keyedObject, *chain.CertificateChainData)
+}
+
+var (
+	objectOperatorsMap = map[objectKind]objectOperators{
+		mutatingWebhookType: {
+			creator:         initMutatingWebhook,
+			toChainMapper:   mapWebhookToChain,
+			fromChainMapper: mapWebhookFromChain,
+		},
+		validatingWebhookType: {
+			creator:         initValidatingWebhook,
+			toChainMapper:   mapWebhookToChain,
+			fromChainMapper: mapWebhookFromChain,
+		},
+		secretType: {
+			creator:         initSecret,
+			toChainMapper:   mapSecretToChain,
+			fromChainMapper: mapSecretFromChain,
+		},
 	}
-	return clientConfigList
+)
+
+// readCertificateChain is the entry point to read from K8s all the certificate
+// chain data related to the managed webhooks. This information is both tracked
+// as a K8s object map and a certificate chain specific structure which should
+// be supplied to this method initialized and empty.
+func (m *Manager) readCertificateChain(objects objectMap, certificateChain *chain.CertificateChainData) error {
+	m.initObjects(objects)
+	certificateChain.CA.Name = m.secretCAName().String()
+	err := m.readObjectsToChain(objects, certificateChain)
+	return err
 }
```

Comment:
What do you think about changing this into
```suggestion
func (m *Manager) readCertificateChain(objects objectMap) (*chain.CertificateChain, error) {
	m.initObjects(objects)
	certificateChain := newChain()
	certificateChain.CA.Name = m.secretCAName().String()
	err := m.readObjectsToChain(objects, certificateChain)
	if err != nil {
	    return nil, err
	}
	return certificateChain, nil
}
```

Then we don't need to pass Options to `CertificateChain` we just have a Update Receiver on it that pass the Update options, this way we can remove `certificateChain` (removing also de logs and the timer since it can be set globally at triple.time)